File: src/php/ParseHTML.php
<?php

include("simplehtmldom/simple_html_dom.php");
include("FoodItem.php");
include("PopulateDB.php");
include_once "LogFileInfo.php";

$DAY_LIST = array(
    "Monday",
    "Tuesday",
    "Wednesday",
    "Thursday",
    "Friday",
    "Saturday",
    "Sunday",
);

$MEAL_TYPE = array("Breakfast", "Lunch", "Dinner");

parseHTML($DAY_LIST, $MEAL_TYPE);

function parseHTML($DAY_LIST, $MEAL_TYPE) {
    //make sure directory exist
    checkDir();
    $file = openLogFile(PARSE_LOG_FILE_NAME, "COMMONS MENU PARSE LOG FILE");
    writeToFile($file, "\n===================================\t
        Begin new Log\t===================================");
    writeToFile($file, "ParseHTML() :: parseHTML()");

    $url = "http://dining.kennesawstateauxiliary.com/commonsmenu.htm";
    writeToFile($file, "ParseHTML() :: Parsing $url");
    $document = file_get_html($url);
    $body = $document->getElementByTagName("body");

    prepareDataRetrieval($body, $DAY_LIST, $MEAL_TYPE, $file);

    unset($body);
    unset($document);
}

function prepareDataRetrieval(simple_html_dom_node $body, $DAY_LIST, $MEAL_TYPE, $file) {
    writeToFile($file, "ParseHTML() :: prepareDataRetrieval()");

    /*create array dictionary for food items
        [itemname : FoodItem]
    */
    $foodItemMap = array();
    $currentMeal = "";  //will store the current meal html traversal below
    $currentStation = "";   //will store the current station

    writeToFile($file, "ParseHTML() :: Pulling date of week");
    //get current week
    $weekOf = $body->find("td[class=titlecell]", 0);
    $weekOf = $weekOf->find("span", $weekOf->size-1)->innertext;

    /*pull information for each day of the week*/
    foreach ($DAY_LIST as &$day) {
        /*use $day variable to retrieve html tag with class name of $day*/
        $dayInnerTable = $body->getElementById(strtolower($day))->find("table[class=dayinner]", 0);
        if(!isset($dayInnerTable)) {
            continue;
        }
        //pull information for each meal of the day
        foreach ($MEAL_TYPE as &$mealTime) {
            $meal = strtolower(substr($mealTime,0,3));

            $tableRows = $dayInnerTable->find("tr[class=$meal]");
            if (count($tableRows) < 1) {
                if ($tableRows != null) {
                    unset($tableRows);
                }
                continue;
            }

            foreach ($tableRows as &$row) {

                $firstChild = $row->first_child();
                $className = $firstChild->getAttribute("class");
                $dataStr = removeWhiteSpace($firstChild->text());

                if($className == "mealname") {
                    if (strlen($dataStr) > 0) {
                        $currentMeal = $dataStr;
                    }
                } else if($className == "station") {
                    if (strlen($dataStr) > 0) {
                        $currentStation = $dataStr;
                    }

                    /*check if the row contains the venue station... validate this
                    by removing unwanted characters in the string and checking the
                    remaining string length*/

                    /*the current system KSU uses has a <td class="menuitem">
                    * which holds the food item name, located under a <td> tag
                    * with a station class name. However, only the first instance
                    * of the current station will be listed. Each other subsequent
                    * items associated with this station will be located under an
                    * empty td tag of class name "station" */

                    $fooditem = pullFoodItemNames($row, $currentMeal, $currentStation, $day, $weekOf);
                    //store food item in array dictionary
                    if(isset($fooditem)) {
                        $foodItemMap [string_sanitize(removeWhiteSpace($fooditem->getItem_name()), 2)] = $fooditem;
                    }
                }

                unset($firstChild);    //free memory
                unset($dataStr);
                unset($className);


            }
            unset($row);    //deallocate $row
            unset($tableRows);
        }
        unset($mealTime);   //deallocate $meal
    }
    unset($day);    //deallocate $day

    parseNutritionData($foodItemMap, $body, $file);

    writeToFile($file, "ParseHTML() :: Starting PopulateDB()");
    writeToFile($file, "\n===================================\t
        End Log\t===================================");
    //start DB populating
    startDBAccess($foodItemMap);
    closeFile($file);
}

/*
 * removes white spaces of various types from string */
function removeWhiteSpace($str) {
    $str = str_replace("&nbsp;", "", $str);
    $str = str_replace(" ", "", $str);
    return trim($str);
}

/*creates a FoodItem() object and sets attributes to the data passed in method
* method parameters
* */
function pullFoodItemNames(simple_html_dom_node $row, $mealStr, $stationStr, $dayStr, $weekOfStr) {
    echo "Item Created<br>";
    $foodItem = new FoodItem();
    $element = $row->find("span", 0);
    if(!isset($element)) {
        $element = $row->find("[class=nonuts]", 0);
        if(!isset($element)) {
            return null;
        }
    }

    $foodItem->setItem_name($element->convert_text($element->text()));
    $foodItem->setItem_price(removeWhiteSpace($row->find("[class=price]", 0)->text()));
    $foodItem->setStation(string_sanitize($stationStr, 0));
    $foodItem->setDayname($dayStr);
    $foodItem->setDay(getDayNum($dayStr));
    $foodItem->setMeal($mealStr[0] . strtolower(substr($mealStr, 1)));
    $foodItem->setMenudate($weekOfStr);

    return $foodItem;
}

function parseNutritionData(array $foodItemMap, simple_html_dom_node $body, $file){
    $scripts = $body->find("script");
    $parentClass = "nutrPkey";
    $childClass = "nutrDat";
    $newHtml = "";

    /*iterate through each script and append that html to $newHtml
    ...this html will be parsed and used later
    */
    foreach ($scripts as &$scr) {
        $newHtml .= $scr->innertext();
    }
    unset ($scr);
    unset($scripts);
    /*The following is to get the information from the javascript array*/

    //create closing and opening div tag for parent. text is pkey
    //===========================================================
    /* regex find: \saData\[\'
     * replace: </div>\n<div class="nutPkey">
    */
    $pat = "/\\saData\\[\'/";
    $newHtml = preg_replace($pat, "</div><div class=\"$parentClass\"><div class='firstChild'>", $newHtml);

    /* * create child div tag for data
      * ====================================
      * regex find: \']=[A-z]+ [A-z]+\(
      * replace: \n<div class="nutItemData">
    */
    $pat = "/\']=[A-z]+ [A-z]+\\(/";
    $newHtml = preg_replace($pat, "</div><div class=\"$childClass\">", $newHtml);

    /* close child div
    * ===================================
    * regex find: \);
    * replace: \n</div>
     * note: first pass has (?!.*\\);) added to pattern to replace all matches
     * except the last occurrence
    * */
    $pat = "/\\);(?!.*\\);)/";
    $newHtml = preg_replace($pat, "\n</div>\n", $newHtml);
    /*one more pass to close all relevant div tags*/
    $pat = "/\\);/";
    $newHtml = preg_replace($pat, "\n</div></div>\n", $newHtml);

    /*remove all comments starters..not the entire comment block*/
    $newHtml = str_replace("<!--", "", $newHtml);

    //create new body element from new html
    $body = str_get_html($newHtml);
    unset($newHtml); //free memory

    $parents = $body->find("[class=$parentClass]");

    if(isset($parents)) {
        writeToFile($file, "ParseHTML() :: Begin storing Food Data");

        foreach ($parents as $parent) {
            /*replace quotes used for each data item.. split string*/
            $data = preg_split("/,,,/", str_replace("'", ",", $parent->find("[class=$childClass]", 0)->text()));

            /*remove unwanted characters from first and last index*/
            $data[0] = str_replace(",", "", $data[0]);
            $data[count($data) - 1] = removeWhiteSpace($data[count($data) - 1]);
            $data[count($data) - 1] = removeWhiteSpace(str_replace(",", "", $data[count($data) - 1]));

            $nameKey = string_sanitize(removeWhiteSpace($data[22]), 2);
            try {
                storeFoodData($foodItemMap[$nameKey], $data, $parent->find("[class=firstChild]", 0)->text(), $nameKey, $file);

            } catch (Exception $e){
                $e->getTrace();
            }
            unset($nameKey);
            unset($data);
        }
        unset($parent);
    }
    writeToFile($file, "ParseHTML() :: Parsing Complete");

}

function storeFoodData($foodItem, array $data, $pkey, $name, $file) {
    $nvar = null;

    if(isset($foodItem)) {

//        $foodItem->setPkey(sanitizeInt($pkey));
        //serving size
        $foodItem->setServ_size($data[0]);

        //Calories
        $nvar = intval($data[1], 10);
        $foodItem->setCalories("".(int)(($nvar < 5) ? 0 : (($nvar >= 5 && $nvar <= 50) ? round($nvar/5, 0)*5 : round($nvar/10, 0)*10)));

        //fat calories
        $nvar = intval($data[2]);
        $cf_pct = (!$foodItem->getCalories() == 0) ? round(($nvar/$foodItem->getCalories())*100, 0) : 0;
        $foodItem->setCalfat("".(int)(($nvar < 5) ? 0 : (($nvar >= 5 && $nvar <= 50) ? round($nvar/5, 0)*5 : round($nvar/10, 0)*10))." ($cf_pct %)");

        //fat
        $nvar = round(floatval($data[3]), 2);
        $foodItem->setFat("". (($nvar<.50) ? 0 : (($nvar>=.50 && $nvar<5.00) ? round(round($nvar/.5,0)*.5,1) : round($nvar,0)))."g");

        //fat percent
        $foodItem->setFat_pct_dv("". intval($data[4])."%");

        //sat fat
        $nvar=round(floatval($data[5]),2);
        $foodItem->setSatfat("". (($nvar < .50) ? 0 : (($nvar >= .50 && $nvar < 5.00) ? round(round($nvar / .5, 0) * .5, 1) : round($nvar, 0)))."g");
        //sat fat percent
        $foodItem->setSatfat_pct_dv("".intval($data[6])."%");

        //transfat
        $nvar=round(floatval($data[7]),2);
        $foodItem->setTransfat("". (($nvar<.50) ? 0 : (($nvar>=.50 && $nvar<5.00) ? round(round($nvar/.5,0)*.5,1) : round($nvar,0)))."g");
        //cholesterol
        $nvar = intval($data[8]);
        $foodItem->setChol("". (($nvar<2) ? 0 : (($nvar>=2 && $nvar<5) ? "< 5" : (int)round($nvar/5,0)*5))."mg");

        //chol percent
        $foodItem->setChol_pct_dv("". intval($data[9])."%");

        //sodium
        $nvar= intval($data[10]);
        $foodItem->setSodium("".(int)(($nvar<5) ? 0 : (($nvar>=5 && $nvar<=140) ? round($nvar/5,0)*5 : round($nvar/10,0)*10))."mg");
        //sodium %
        $foodItem->setSodium_pct_dv("".intval($data[11])."%");

        //carbo
        $nvar=round(floatval($data[12]), 2);
        $foodItem->setCarbo("". (($nvar<.50) ? "0" : (($nvar>=.50 && $nvar<1.0) ? "< 1" : (int)round($nvar,0)))."g");
        //carbo %
        $foodItem->setCarbo_pct_dv("". intval($data[13])."%");

        //dietary fiber
        $nvar=round(floatval($data[14]), 2);
        $foodItem->setDfib("". (($nvar<.50) ? "0" : (($nvar>=.50 && $nvar<1.0) ? "< 1" : (int)round($nvar,0)))."g");
        //dietary fib %
        $foodItem->setDfib_pct_dv("". intval($data[15])."%");

        //sugars
        $nvar=round(floatval($data[16]), 2);
        $foodItem->setSugars("". (($nvar<.50) ? "0" : (($nvar>=.50 && $nvar<1.0) ? "< 1" : (int)round($nvar,0)))."g");

        //protein
        $nvar=round(floatval($data[17]), 2);
        $foodItem->setProtein("". (($nvar<.50) ? 0 : (($nvar>=.50 && $nvar<1.00) ? 1  : (int)round($nvar,0)))."g");

        //vit a (compare to $data[18])
        $nvar=round(floatval($data[25])/5000 * 100, 1);
        $foodItem->setVita_pct_dv("". (int)(($nvar<1.0) ? 0 : (($nvar<=2) ? 2 : (($nvar<=10.0) ? round(round($nvar/2,1)*2,0) : (($nvar<=50) ? round(round($nvar/5,1)*5,0) : round($nvar/10,0)*10))))."%");

        //vit c (compare to $data[19])
        $nvar=round(floatval($data[26])/60 * 100, 1);
        $foodItem->setVitc_pct_dv("". (int)(($nvar<1.0) ? 0 : (($nvar<=2) ? 2 : (($nvar<=10.0) ? round(round($nvar/2,1)*2,0) : (($nvar<=50) ? round(round($nvar/5,1)*5,0) : round($nvar/10,0)*10))))."%");

        //calcium (compare to $data[20])
        $nvar=round(floatval($data[27])/1000 * 100, 1);
        $foodItem->setCalcium_pct_dv("". (int)(($nvar<1.0) ? 0 : (($nvar<=2) ? 2 : (($nvar<=10.0) ? round(round($nvar/2,1)*2,0) : (($nvar<=50) ? round(round($nvar/5,1)*5,0) : round($nvar/10,0)*10))))."%");

        //iron (compare to $data[21])
        $nvar=round(floatval($data[28])/18 * 100, 1);
        $foodItem->setIron_pct_dv("".(int)(($nvar<1.0) ? 0 : (($nvar<=2) ? 2 : (($nvar<=10.0) ? round(round($nvar/2,1)*2,0) : (($nvar<=50) ? round(round($nvar/5,1)*5,0) : round($nvar/10,0)*10))))."%");

        //$data[22] is $foodItem named.. already stored...
        $foodItem->setItem_desc($data[23]);
        $allergArr = preg_split("/,/",$data[24]);
        $al = "";
        foreach($allergArr as &$allergen) {
            $temp = string_sanitize($allergen, 2);
            $temp = str_replace("Contains", "", $temp);
            if($temp == "MSG") {
                $al .= $allergen.",";
            } else {
                $al .= strtolower(substr($temp, 0, 3)).",";
            }
        }
        unset($allergen);
        unset($allergArr);
        $foodItem->setAllergens($al);
        $foodItem->setVeg_type($data[29]);

    } else {
        writeToFile($file, "Failed at: $name");
    }
}

function string_sanitize($str, $quoteStyle) {
    $ENT = null;
    if($quoteStyle == 0) {
        $ENT = ENT_NOQUOTES;
    } else if ($quoteStyle == 1) {
        $ENT = ENT_COMPAT;
    } else {
        $ENT = ENT_QUOTES;
    }

    $str = str_replace("&nbsp;", "", $str);
    $str = preg_replace("/[^a-zA-Z0-9]+/", "", html_entity_decode($str, $ENT));
    return $str;
}

function sanitizeInt($str) {
    return preg_replace("/[^0-9]+/", "", html_entity_decode($str, ENT_QUOTES));
}
function getDayNum($dayStr) {
    switch($dayStr) {
        case "Monday":
            return "1";
        case "Tuesday":
            return "2";
        case "Wednesday":
            return "3";
        case "Thursday":
            return "4";
        case "Friday":
            return "5";
        case "Saturday":
            return "6";
        case "Sunday":
            return "7";
        default:
            // we shouldn't get here
            return "1";
    }
}

File: src/java/ParseHtml.java
package java;
import org.jsoup.Connection;
import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

import java.io.*;
import java.text.DecimalFormat;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.regex.Matcher;
import java.util.regex.Pattern;


public class ParseHtml {

    private static final ArrayList<String> DAY_LIST = new ArrayList<String>();

    private static final String[] COURSE_LIST = {"Breakfast", "Lunch", "Dinner"};

    static {
        DAY_LIST.add("Monday");
        DAY_LIST.add("Tuesday");
        DAY_LIST.add("Wednesday");
        DAY_LIST.add("Thursday");
        DAY_LIST.add("Friday");
        DAY_LIST.add("Saturday");
        DAY_LIST.add("Sunday");
    }

    public static void parseHtml(String url) throws IOException {

        /*establish connection to url*/
//        Connection con = Jsoup.connect(url);
        /*get html string from url*/
//        String html = con.get().html();

        String fileName = "Menu.html";
        String dirPath = "html";
        File dir = new File(dirPath);
        if(!dir.exists()) {
            dir.mkdir();
        }

        File file = new File(dir, fileName);

        StringBuilder htmlString = new StringBuilder();
        char[] buf = new char[1024];

        FileInputStream inputStream = new FileInputStream(file);
        InputStreamReader reader = new InputStreamReader(inputStream, "utf-8");

        while(reader.read(buf) > -1) {
            htmlString.append(buf);
        }

        /*parse html string into html document*/
        Document doc = Jsoup.parse(htmlString.toString());
        /*retrieve body tag from document*/
        Element body = doc.body();

        prepareDataRetrieval(body);
    }

    private static void prepareDataRetrieval(Element body) {
        /*create*/
//        ArrayList<FoodItem> foodList = new ArrayList<>();

        HashMap <String, FoodItem> foodItemMap = new HashMap<>();
        String currentMeal = "";
        String currentStation = "";
        String weekOf = body.getElementsByClass("titlecell").first().getElementsByTag("span").last().text();

        System.out.println("Week of: " + weekOf);
        /*pull data for each day*/
        for (String day : DAY_LIST) {
            /*find id in html document as a lowercase string*/
            Element dayTag = body.getElementById(day.toLowerCase());
            try {

                Element dayinnerTable = dayTag.getElementsByClass("dayinner").first();

                for (String mealTime : COURSE_LIST) {
                    String meal = mealTime.substring(0, 3).toLowerCase();

                    Elements tableRows = dayinnerTable.getElementsByClass(meal);

                    for (Element row : tableRows) {
                        Element firstChild = row.child(0);

                        if (firstChild.className().equals("mealname")) {
                    /*if the first child is the meal course name, store
                     * that name in the corresponding variable*/
                            currentMeal = firstChild.text();
                        } else if (firstChild.className().equals("station")) {
                    /*if the first child is the venue station, store
                    * that name in the corresponding variable*/

                        /*get rid of any unwanted space characters*/
                            String tempStation = firstChild.text().replace("\u00a0","");
                            if (tempStation.trim().length() > 0) {
                                currentStation = tempStation;
                            }

                        /*the current system KSU uses has a <td class="menuitem">
                        * which holds the food item name, located under a <td> tag
                        * with a station class name. However, only the first instance
                        * of the current station will be listed. Each other subsequent
                        * items associated with this station will be located under an
                        * empty td tag of class name "station" */

                            FoodItem item = pullFoodItemNames(row, currentMeal, currentStation, day, weekOf);
                            foodItemMap.put(item.getItem_name().replaceAll(" ", ""), item);


                        }

                    }
                }


            } catch (NullPointerException e) {
                e.printStackTrace();
            }

        }
        parseNutritionData(foodItemMap, body);
    }

    private static FoodItem pullFoodItemNames(Element row, String meal, String station, String day, String weekOf) {
        FoodItem item = new FoodItem();
        Element element;
        String price;

        if (row.getAllElements().hasClass("nonuts")) {
            /*for items with no nutritional information on webpage*/
            element = row.getElementsByClass("nonuts").first();
        } else {
            element = row.getElementsByTag("span").first();
        }

        /*pull and set name text from element*/
        String foodItemName = element.ownText();
        price = row.getElementsByClass("price").first().text();

        item.setItem_name(foodItemName);
        item.setItem_price(price);
        item.setStation(station);
        item.setDayname(day);
        item.setDay("" + (DAY_LIST.indexOf(day) + 1));
        item.setMeal(meal.charAt(0) + meal.substring(1).toLowerCase());
        item.setMenudate(weekOf);

        return item;
    }

    private static void parseNutritionData(HashMap<String, FoodItem> foodItemMap, Element body) {
        Elements scripts = body.select("script");
        String parentClass = "nutPkey";
        String childClass = "nutItemData";

        /*get all script text and store in string.*/
        StringBuilder sb = new StringBuilder();
        for (Element s : scripts) {
            sb.append(s.html());
        }

        //create closing and opening div tag for parent. text is pkey
        //===========================================================
        /* regex find: \saData\[\'
         * replace: </div>\n<div class="nutPkey">
        */
        Pattern pat = Pattern.compile("\\saData\\[\'");
        String html = pat.matcher(sb.toString()).replaceAll("</div>\n<div class=\"" + parentClass + "\">");

        /* * create child div tag for data
          * ====================================
          * regex find: \']=[A-z]+ [A-z]+\(
          * replace: \n<div class="nutItemData">
        */
        pat = Pattern.compile("\']=[A-z]+ [A-z]+\\(");
        html = pat.matcher(html).replaceAll("\n<div class=\"" + childClass + "\">");

        /* close child div
         * ===================================
         * regex find: \);
         * replace: \n</div>
         * */
        pat = Pattern.compile("\\);");
        html = pat.matcher(html).replaceAll("\n</div>\n");

        /*remove all comments starters..not the entire comment block*/
        html = html.replaceAll("<!--", "");

        //create new body element from new html
        body = Jsoup.parse(html).body();

        sb.delete(0, sb.length());
        Elements parents = body.getElementsByClass(parentClass);
        System.out.println(parents.size());
        for (Element parent : parents) {
            String pkey;

            /*last element usually has added, unwanted text.. so using a regex
            * pattern, store as the pkey, only the wanted information*/
            if (parent == parents.last()) {
                pat = Pattern.compile("[a-zA-z0-9]+_[a-zA-z0-9]+");
                Matcher matcher = pat.matcher(parent.ownText());

                pkey = (matcher.find()) ? matcher.group(0) : parent.ownText();

//                Turkey Baguette with Chipotle Mayo
            } else {
                pkey = parent.ownText();

            }

            /*replace quotes used for each data item.. split string*/
            String[] data = parent.child(0).text().replace("\'", ",").split(",,,");

            /*remove unwanted characters from first and last index*/
            data[0] = data[0].substring(1);
            data[data.length-1] = data[data.length-1].substring(0, data[data.length-1].length()-1);
            data[data.length-1] = data[data.length-1].trim();

            /*store data in food item from hash map*/
            /*data[22] is item name*/
            String name = data[22].replaceAll(" ", "");
            storeFoodData(foodItemMap.get(name),data, pkey, name);

        }

    }

    private static void storeFoodData(FoodItem item, String[] data, String pkey, String name) {
        double nvar;

        if (item != null) {
            DecimalFormat format = new DecimalFormat();

            item.setPkey(pkey);
            //serving size
            item.setServ_size(data[0]);

            //Calories
            nvar = Integer.parseInt(data[1]);
            item.setCalories(""+(int)((nvar < 5) ? 0 : ((nvar >= 5 && nvar <= 50) ? round(nvar/5, 0)*5 : round(nvar/10, 0)*10)));

            //fat calories
            nvar = Integer.parseInt(data[2]);
            item.setCalfat(""+(int)((nvar < 5) ? 0 : ((nvar >= 5 && nvar <= 50) ? round(nvar/5, 0)*5 : round(nvar/10, 0)*10)));

            //fat
            nvar = round(Float.parseFloat(data[3]), 2);
            item.setFat(""+ format.format((nvar<.50) ? 0 : ((nvar>=.50 && nvar<5.00) ? round(round(nvar/.5,0)*.5,1) : round(nvar,0))));

            //fat percent
            item.setFat_pct_dv("" + Integer.parseInt(data[4]));

            //sat fat
            nvar=round(Float.parseFloat(data[5]),2);
            item.setSatfat(""+ format.format((nvar < .50) ? 0 : ((nvar >= .50 && nvar < 5.00) ? round(round(nvar / .5, 0) * .5, 1) : round(nvar, 0))));
            //sat fat percent
            item.setSatfat_pct_dv(""+Integer.parseInt(data[6]));

            //transfat
            nvar=round(Float.parseFloat(data[7]),2);
            item.setTransfat(""+ format.format((nvar<.50) ? 0 : ((nvar>=.50 && nvar<5.00) ? round(round(nvar/.5,0)*.5,1) : round(nvar,0))));
            //cholesterol
            nvar = Integer.parseInt(data[8]);
            item.setChol("" + ((nvar<2) ? 0 : ((nvar>=2 && nvar<5) ? "< 5" : (int)round(nvar/5,0)*5)));

            //chol percent
            item.setChol_pct_dv("" + Integer.parseInt(data[9]));

            //sodium
            nvar= Integer.parseInt(data[10]);
            item.setSodium(""+(int)((nvar<5) ? 0 : ((nvar>=5 && nvar<=140) ? round(nvar/5,0)*5 : round(nvar/10,0)*10)));
            //sodium %
            item.setSodium_pct_dv(""+Integer.parseInt(data[11]));

            //carbo
            nvar=round(Float.parseFloat(data[12]), 2);
            item.setCarbo("" + ((nvar<.50) ? "0" : ((nvar>=.50 && nvar<1.0) ? "< 1" : (int)round(nvar,0))));
            //carbo %
            item.setCarbo_pct_dv("" + Integer.parseInt(data[13]));

            //dietary fiber
            nvar=round(Float.parseFloat(data[14]), 2);
            item.setDfib("" + ((nvar<.50) ? "0" : ((nvar>=.50 && nvar<1.0) ? "< 1" : (int)round(nvar,0))));
            //dietary fib %
            item.setDfib_pct_dv("" + Integer.parseInt(data[15]));

            //sugars
            nvar=round(Float.parseFloat(data[16]), 2);
            item.setSugars(""+ ((nvar<.50) ? "0" : ((nvar>=.50 && nvar<1.0) ? "< 1" : (int)round(nvar,0))));

            //protein
            nvar=round(Float.parseFloat(data[17]), 2);
            item.setProtein("" + ((nvar<.50) ? 0 : ((nvar>=.50 && nvar<1.00) ? 1  : (int)round(nvar,0))));

            //vit a (compare to data[18])
            nvar=round(Float.parseFloat(data[25])/5000 * 100, 1);
            item.setVita_pct_dv("" + (int)((nvar<1.0) ? 0 : ((nvar<=2) ? 2 : ((nvar<=10.0) ? round(round(nvar/2,1)*2,0) : ((nvar<=50) ? round(round(nvar/5,1)*5,0) : round(nvar/10,0)*10)))));

            //vit c (compare to data[19])
            nvar=round(Float.parseFloat(data[26])/60 * 100, 1);
            item.setVitc_pct_dv("" + (int)((nvar<1.0) ? 0 : ((nvar<=2) ? 2 : ((nvar<=10.0) ? round(round(nvar/2,1)*2,0) : ((nvar<=50) ? round(round(nvar/5,1)*5,0) : round(nvar/10,0)*10)))));

            //calcium (compare to data[20])
            nvar=round(Float.parseFloat(data[27])/1000 * 100, 1);
            item.setCalcium_pct_dv("" + (int)((nvar<1.0) ? 0 : ((nvar<=2) ? 2 : ((nvar<=10.0) ? round(round(nvar/2,1)*2,0) : ((nvar<=50) ? round(round(nvar/5,1)*5,0) : round(nvar/10,0)*10)))));

            //iron (compare to data[21])
            nvar=round(Float.parseFloat(data[28])/18 * 100, 1);
            item.setIron_pct_dv(""+(int)((nvar<1.0) ? 0 : ((nvar<=2) ? 2 : ((nvar<=10.0) ? round(round(nvar/2,1)*2,0) : ((nvar<=50) ? round(round(nvar/5,1)*5,0) : round(nvar/10,0)*10)))));

            //data[22] is item named.. already stored...
            item.setItem_desc(data[23]);
            item.setAllergens(data[24]);
            item.setVeg_type(data[29]);

            System.out.println(item.getItem_name());
            System.out.println(item.getPkey());
            System.out.println(item.getItem_desc());
            System.out.println(item.getCalories());
            System.out.println(item.getCalfat());
            System.out.println(item.getFat());
            System.out.println(item.getFat_pct_dv());
            System.out.println(item.getSatfat());
            System.out.println(item.getSatfat_pct_dv());
            System.out.println(item.getTransfat());
            System.out.println(item.getChol());
            System.out.println(item.getChol_pct_dv());
            System.out.println(item.getSodium());
            System.out.println(item.getSodium_pct_dv());
            System.out.println(item.getCarbo());
            System.out.println(item.getCarbo_pct_dv());
            System.out.println(item.getDfib());
            System.out.println(item.getDfib_pct_dv());
            System.out.println(item.getSugars());
            System.out.println(item.getProtein());
            System.out.println(item.getVita_pct_dv());
            System.out.println(item.getVitc_pct_dv());
            System.out.println(item.getCalcium_pct_dv());
            System.out.println(item.getIron_pct_dv());
            System.out.println(item.getAllergens());

            System.out.println(item.getMeal());
            System.out.println(item.getStation());
            System.out.println(item.getMenudate());
            System.out.println(item.getServ_size());
            System.out.println(item.getVeg_type());
            System.out.println(item.getDay());
            System.out.println(item.getDayname());
            System.out.println("========================================\n\n");
        } else {
            System.out.println("failed at: " + pkey + "\n for:\n" + name);
        }

    }

    public static void outPutHTML (String url) throws IOException {
        Connection con = Jsoup.connect(url);

        /*the following is a test to see if the outer html and inner html methods retreive similar or different results.*/
        String outerHtml = con.get().outerHtml();
        String innerHtml = con.get().html();
        String innerName = "innerHTML.txt";
        String outerName = "outerHTML.txt";

        String innerDesc = "Using Jsoup .connect(url).get().html to retrieve the inner HTML of a web page"
                + "\n============================================================\n\n\n";
        String outerDesc = "Using Jsoup .connect(url).get().html to retrieve the outer HTML of a web page"
                + "\n============================================================\n\n\n";

        outPutToFile(innerName, innerHtml, innerDesc);
        outPutToFile(outerName, outerHtml, outerDesc);
    }

    public static void outPutToFile(String fileName, String data, String desc) throws IOException{
        String methodName = "outPutToFile";

        String dirPath = "output";
        File dir = new File(dirPath);
        if(!dir.exists()) {
            dir.mkdir();
        }

        File file = new File(dir, fileName);

        /*if the file does not already exist then create a new file*/
        if (!file.exists()) {
            printLog(methodName, "Creating File");
            file.createNewFile();
        }

        FileOutputStream fileOutputStream = new FileOutputStream(file);
        OutputStreamWriter outputStreamWriter = new OutputStreamWriter(fileOutputStream, "utf-8");
        printLog(methodName, "Writing File");
        outputStreamWriter.write(desc);
        outputStreamWriter.write(data);

        printLog(methodName, "Closing Output Streams");
        outputStreamWriter.close();
        fileOutputStream.close();
        printLog(methodName, "Output Complete");
    }

    private static void printLog (String method, String text) {
        System.out.println(method + "\t:\t" + text);
    }


    /*x is the number of places to round*/
    private static double round(int nVal, int x) {
        return Math.round(nVal * Math.pow(10, x))/Math.pow(10, x);
    }
    private static double round(double nVal, int x) {
        return Math.round(nVal * Math.pow(10, x))/Math.pow(10, x);
    }
}
